clippy_lints/src/path_from_format.rs

```diff
@@ -0,0 +1,136 @@
+                            || string_parts[n+1].starts_with('/')
+                            || string_parts[n+1].starts_with('\\')
+                        ){
+                        span_lint_and_note(
+                            cx,
+                            PATH_FROM_FORMAT,
+                            expr.span,
+                            "`format!(..)` used to form `PathBuf`",
+                            None,
+                            "if it fits your use case, you may want to consider using `Path::new()` and `.join()` to make it OS-agnostic and improve code readability",
+                        );
+                        return;
+                    }
+                    if n == 0 {
+                        if string_parts[0].is_empty() {
+                            sugg = format!("Path::new({})", real_vars[order_of_real_vars[0]]);
+                        }
+                        else {
+                            push_comps(&mut sugg, Path::new(string_parts[0]));
+                            let _ = write!(sugg, ".join({})", real_vars[order_of_real_vars[0]]);
+                        }
+                    }
+                    else if string_parts[n].is_empty() {
+                            sugg = format!("{sugg}.join({})", real_vars[order_of_real_vars[n]]);
+                    }
+                    else {
+                        let mut string = String::from(string_parts[n]);
+                        if string.starts_with('/') || string.starts_with('\\') {
+                            string.remove(0);
+                        }
+                        push_comps(&mut sugg, Path::new(&string));
+                        let _ = write!(sugg, ".join({})", real_vars[order_of_real_vars[n]]);
+                    }
+                }
+                if !string_parts[real_vars.len()].is_empty() {
+                    let mut string = String::from(string_parts[real_vars.len()]);
+                    if string.starts_with('/') || string.starts_with('\\') {
+                        string.remove(0);
+                    }
+                    push_comps(&mut sugg, Path::new(&string));
+                }
+                span_lint_and_sugg(
+                    cx,
+                    PATH_FROM_FORMAT,
+                    expr.span,
+                    "`format!(..)` used to form `PathBuf`",
+                    "consider using `Path::new()` and `.join()` to make it OS-agnostic and improve code readability",
+                    sugg,
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        }
+    }
+}
+
+fn push_comps(string: &mut String, path: &Path) {
+    let comps = path.components();
+    for n in comps {
+        let x = n.as_os_str().to_string_lossy().to_string();
+        if string.is_empty() {
+            let _ = write!(string, "Path::new(\"{x}\")");
+        } else {
+            let _ = write!(string, ".join(\"{x}\")");
```

Comment:
Any reason why we're using `write!` macro, instead of using `format!`?

Reply:
iirc, it was because using `format!` was not recommended by another clippy lint